src/main/java/uk/gov/hmcts/darts/courthouse/service/CourthouseServiceImpl.java

```diff
@@ -48,4 +52,43 @@ public Courthouse addCourtHouse(uk.gov.hmcts.darts.courthouse.model.Courthouse c
         Courthouse mappedEntity = this.mapper.mapToEntity(courthouse);
         return repository.saveAndFlush(mappedEntity);
     }
+
+
+    /**
+     * retrieves the courtroom from the database.
+     *
+     * @param courthouseCode Optional parameter. If it is not provided, then name will be used by itself.
+     * @param courthouseName Name of the courthouse to search for.
+     * @return the found courtroom
+     * @throws CourthouseNameNotFoundException when the courthouse isn't found
+     * @throws CourthouseCodeNotMatchException when the courtroom is found, but it has a different code that expected.
+     */
+    @Override
+    public Courthouse retrieveCourtHouse(Integer courthouseCode, String courthouseName)
+        throws CourthouseNameNotFoundException, CourthouseCodeNotMatchException {
+        String courthouseNameUC = StringUtils.upperCase(courthouseName);
+        Optional<Courthouse> courthouseOptional = Optional.empty();
+        if (courthouseCode != null) {
+            courthouseOptional = repository.findByCode(courthouseCode.shortValue());
+        }
+        if (courthouseOptional.isEmpty()) {
+            //code not found, lookup name instead
+            courthouseOptional = repository.findByCourthouseName(courthouseNameUC);
+            if (courthouseOptional.isEmpty()) {
+                throw new CourthouseNameNotFoundException(courthouseNameUC);
+            }
+            Courthouse courthouse = courthouseOptional.get();
+            if (courthouse.getCode() == null && courthouseCode != null) {
+                //update courthouse with new code
+                courthouse.setCode(courthouseCode.shortValue());
```

Comment:
I guess the assumption here is that when this method is called the courthouse code is valid and is actually for the courthouseName.

Reply:
i've decided to rewrite this section